msal/src/Microsoft.Identity.Client/DeviceCodeResult.cs

```diff
@@ -0,0 +1,112 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.Identity.Client
+{
+    /// <summary>
+    /// This object is returned as part of the device code flow
+    /// and has information intended to be shown to the user about
+    /// where to navigate to login and what the device code is that needs
+    /// to be entered.
+    /// </summary>
+    public class DeviceCodeResult
+    {
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="userCode"></param>
+        /// <param name="deviceCode"></param>
+        /// <param name="verificationUrl"></param>
+        /// <param name="expiresOn"></param>
+        /// <param name="interval"></param>
+        /// <param name="message"></param>
+        /// <param name="clientId"></param>
+        /// <param name="resource"></param>
+        public DeviceCodeResult(
+            string userCode,
+            string deviceCode,
+            string verificationUrl,
+            DateTimeOffset expiresOn,
+            long interval,
+            string message,
+            string clientId,
+            string resource)
+        {
+            UserCode = userCode;
+            DeviceCode = deviceCode;
+            VerificationUrl = verificationUrl;
+            ExpiresOn = expiresOn;
+            Interval = interval;
+            Message = message;
+            ClientId = clientId;
+            Resource = resource;
+        }
+
+        /// <summary>
+        /// User code returned by the service
+        /// </summary>
+        public string UserCode { get; }
+
+        /// <summary>
+        /// Device code returned by the service
+        /// </summary>
+        public string DeviceCode { get; }
+
+        /// <summary>
+        /// Verification URL where the user must navigate to authenticate using the device code and credentials.
+        /// </summary>
+        public string VerificationUrl { get; }
+
+        /// <summary>
+        /// Time when the device code will expire.
+        /// </summary>
+        public DateTimeOffset ExpiresOn { get; }
+
+        /// <summary>
+        /// Polling interval time to check for completion of authentication flow.
+        /// </summary>
+        public long Interval { get; }
+
+        /// <summary>
+        /// User friendly text response that can be used for display purpose.
+        /// </summary>
+        public string Message { get; }
+
+        /// <summary>
+        /// Identifier of the client requesting device code.
+        /// </summary>
+        public string ClientId { get; }
+
+        /// <summary>
+        /// Identifier of the target resource that would be the recipient of the token.
+        /// </summary>
+        public string Resource { get; }
+        // todo: this should be SCOPE(S) and not resource for v2 endpoint...
+    }
+}

```

Comment:
I know this is public, but looks to be similar to what is in ADAL. Is there a way to move the shared code to Core and surface the public api to ADAL and MSAL respectively?

Reply:
Yes, good point. You can eliminate duplicate code by adding it to core. However, we do not currently have any public code in core, so this warrants a discussion. Please note that core.dll assembly is just a convenience to hold files and run tests against - ADAL and MSAL directly include all *.cs files from core. 

So we could move DeviceCodeResult file into core. We don't want to introduce breaking changes, so the ADAL namespaces cannot change. Therefore, when adding it to core, we have do something like: 

#if ADAL
namespace Microsoft.IdentityModel.Clients.ActiveDirectory
#else 
namespace Microsoft.Identity.Client
#endif

Reply:
Alternatively, if the only commonality are data classes, we could live with the duplication.

Reply:
I actually think the data classes _may_ end up varying since there are protocol changes in the v2 endpoint.  This is why I just made them copies so that we can vary them with the final V2 endpoint versions.